encoding/ccf/decode_typedef.go

```diff
@@ -0,0 +1,477 @@
+
+	case CBORTagContractType:
+		ctr := func(location common.Location, identifier string) cadence.Type {
+			return cadence.NewMeteredContractType(
+				d.gauge,
+				location,
+				identifier,
+				nil,
+				nil,
+			)
+		}
+		return d.decodeCompositeType(types, rawFields, ctr)
+
+	case CBORTagEnumType:
+		ctr := func(location common.Location, identifier string) cadence.Type {
+			return cadence.NewMeteredEnumType(
+				d.gauge,
+				location,
+				identifier,
+				nil,
+				nil,
+				nil,
+			)
+		}
+		return d.decodeCompositeType(types, rawFields, ctr)
+
+	case CBORTagStructInterfaceType:
+		ctr := func(location common.Location, identifier string) cadence.Type {
+			return cadence.NewMeteredStructInterfaceType(
+				d.gauge,
+				location,
+				identifier,
+				nil,
+				nil,
+			)
+		}
+		return d.decodeInterfaceType(types, ctr)
+
+	case CBORTagResourceInterfaceType:
+		ctr := func(location common.Location, identifier string) cadence.Type {
+			return cadence.NewMeteredResourceInterfaceType(
+				d.gauge,
+				location,
+				identifier,
+				nil,
+				nil,
+			)
+		}
+		return d.decodeInterfaceType(types, ctr)
+
+	case CBORTagContractInterfaceType:
+		ctr := func(location common.Location, identifier string) cadence.Type {
+			return cadence.NewMeteredContractInterfaceType(
+				d.gauge,
+				location,
+				identifier,
+				nil,
+				nil,
+			)
+		}
+		return d.decodeInterfaceType(types, ctr)
+
+	default:
+		return ccfTypeID(0),
+			cadenceTypeID(""),
+			fmt.Errorf("unsupported type definition with CBOR tag number %d", tagNum)
+	}
+}
+
+// decodeCompositeType decodes composite type in type definition as
+// language=CDDL
+// composite-type = [
+//
+//	id: id,
+//	cadence-type-id: cadence-type-id,
+//	fields: [
+//	    * [
+//	        field-name: tstr,
+//	        field-type: inline-type
+//	    ]
+//	]
+//
+// ]
+func (d *Decoder) decodeCompositeType(
+	types *cadenceTypeByCCFTypeID,
+	rawFields map[ccfTypeID][]byte,
+	constructor func(common.Location, string) cadence.Type,
+) (ccfTypeID, cadenceTypeID, error) {
+
+	// Decode array head of length 3.
+	err := decodeCBORArrayWithKnownSize(d.dec, 3)
+	if err != nil {
+		return ccfTypeID(0), cadenceTypeID(""), err
+	}
+
+	// element 0: id
+	ccfID, err := d.decodeCCFTypeID()
+	if err != nil {
+		return ccfTypeID(0), cadenceTypeID(""), err
+	}
+
+	// "Valid CCF Encoding Requirements" in CCF specs:
+	//
+	//   "composite-type.id MUST be unique in ccf-typedef-message or ccf-typedef-and-value-message."
+	if types.has(ccfID) {
+		return ccfTypeID(0), cadenceTypeID(""), fmt.Errorf("found duplicate CCF type ID %d in composite-type", ccfID)
+	}
+
+	// element 1: cadence-type-id
+	cadenceID, location, identifier, err := d.decodeCadenceTypeID()
+	if err != nil {
+		return ccfTypeID(0), cadenceTypeID(""), err
+	}
+
+	// element 2: fields
+	rawField, err := d.dec.DecodeRawBytes()
+	if err != nil {
+		return ccfTypeID(0), cadenceTypeID(""), err
+	}
+
+	_ = types.add(ccfID, constructor(location, identifier))
```

Comment:
```suggestion
	// The return value can be ignored, because its non-existence was already checked above
	_ = types.add(ccfID, constructor(location, identifier))
```

Reply:
Done in commit 19f829fd4ca3401237844db355ed247b6321d166.